### PR TITLE
Add compiler flags suggested by security review (#4368)

### PR DESCRIPTION
**Pull request by kf6gpe**
In [4368](https://github.com/flutter/flutter/issues/4368) it's recommended that we build binaries with the following flags that provide security benefits:

- stack cookies to prevent stack overflow (`-fstack_protector_all`)
- read-only relocations (`-Wl,z,relro`),
- `BIND_NOW` to protect the global offset table from being overwritten (`-Wl,-z,now`)
- Position independent executable code generation (`-fPIE`, `-pie`).
- `FORTIFY_SOURCE` to protect from common misuse of memory and string functions. There is an existing comment that FORTIFY_SOURCE appears to have problems with clang, but that's with an unused optimization level (`-O1`) that does not appear to be pertinent to this work. However, I left in the comment should we choose to adjust the optimization levels in the future.

(For a more detailed if slightly dated explanation of what these flags do, see [here](http://developer.blackberry.com/playbook/native/documentation/com.qnx.doc.native_sdk.security/topic/using_compiler_linker_defenses.html).)

Due to size constraints, I did not add the suggested optimization flag (`-O2`) listed in the original issue, per discussions with @goderbauer on the issue.

Changes were applied to the Android binary compilation flags, and relevant changes applied to the iOS compilation flags (stack cookies, offset table and read-only relocations are not supported by the iOS tools). Note that iOS generates code in a position-independent executable way by default, but I explicitly add those directives in a new iOS flag block for future reviewers.

### CODE REVIEW

File: build/config/compiler/BUILD.gn [hunk 1]
```diff
@@ -822,9 +843,10 @@ config("optimize") {
     cflags = [ "-Os" ] + common_optimize_on_cflags  # Favor size over speed.
   } else if (is_android) {
     cflags = [ "-Oz" ] + common_optimize_on_cflags  # Favor size over speed.
-    if (enable_lto) {
-      lto_flags += [ "-flto" ]
-    }
+    # TODO(goderbauer): re-enable when  https://github.com/flutter/flutter/issues/21686 is fixed
```
Comment (goderbauer) [inline build/config/compiler/BUILD.gn]: This looks like a bad merge? I think LTO should stay on.

Comment (kf6gpe) [reply build/config/compiler/BUILD.gn]: Good catch! I think I was too far behind in the tree. Lemme go back and fix that.

**Comment by mit-mit** [comment]
@kf6gpe are we ready to merge this?

**Comment by kf6gpe** [comment]
Oh! Yes, please --- was waiting for an LGTM from @Hixie because it was kind of a sweeping change. Let me resolve the conflicts, and see what he thinks.

**Comment by chinmaygarde** [comment]
-Wa,--noexecstack was reverted in https://github.com/flutter/buildroot/pull/185 due to buildbot failures on Mac. Issue tracked in https://github.com/flutter/flutter/issues/23606.

Also, please format GN file updates using `gn format`.

**Pull request merged**